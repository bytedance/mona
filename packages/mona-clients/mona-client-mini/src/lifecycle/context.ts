import React from 'react';
export type Callback = (...args: any[]) => any;

export enum AppLifecycle {
  launch = 'launch',
  show = 'show',
  hide = 'hide',
  error = 'error',
  pageNotFound = 'pageNotFound',
}

export enum PageLifecycle {
  load = 'load',
  ready = 'ready',
  show = 'show',
  hide = 'hide',
  unload = 'unload',
  resize = 'resize',
  pullDownRefresh = 'pullDownRefresh',
  reachBottom = 'reachBottom',
  shareAppMessage = 'shareAppMessage',
  pageScroll = 'pageScroll',
}

export class LifecycleContext {
  lifecycles: { [key: string]: Callback[] };

  constructor() {
    this.lifecycles = {};
  }

  registerLifecycle(rawName: string, callback: Callback) {
    if (typeof callback !== 'function') {
      return;
    }

    const name = rawName.toLowerCase().replace(/^on/, '');
    this.lifecycles[name] = this.lifecycles[name] || [];
    this.lifecycles[name].push(callback);
  }
}

export const PageLifecycleGlobalContext = React.createContext<LifecycleContext | null>(null);
export const appLifecycleContext = React.createContext<LifecycleContext | null>(null);
