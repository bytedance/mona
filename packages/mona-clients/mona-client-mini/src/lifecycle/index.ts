export { usePageEvent, useAppEvent } from './hooks';
export { PageLifecycleGlobalContext, appLifecycleContext } from './context';
