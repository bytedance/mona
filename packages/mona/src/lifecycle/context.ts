import React from 'react';
export type Callback = (query?: Record<string, string | number | boolean>) => void;

export enum AppLifecycle {
  launch = 'onLaunch',
  show = 'onShow',
  hide = 'onHide',
  error = 'onError',
  pageNotFound = 'onPageNotFound',
}

export enum PageLifecycle {
  load = 'onLoad',
  ready = 'onReady',
  show = 'onShow',
  hide = 'onHide',
  unload = 'onUnload',
  resize = 'onResize',
  pullDownRefresh = 'onPullDownRefresh',
  reachBottom = 'onReachBottom',
  shareAppMessage = 'onShareAppMessage',
  pageScroll = 'onPageScroll',
}

export enum ComponentLifecycle {
  created = 'onCreated',
  attached = 'onAttached',
  ready = 'onReady',
  moved = 'onMoved',
  detached = 'onDetached',
}

export class LifecycleContext {
  lifecycle: Record<string, Set<Callback>>;

  constructor() {
    this.lifecycle = new Proxy<Record<string, Set<Callback>>>(
      {},
      {
        get: function (target, property) {
          if (typeof property === 'string') {
            return target[property.toLowerCase().replace(/^on/, '')];
          }
          return;
        },
        set: function (target, property, value) {
          if (typeof property === 'string') {
            target[property.toLowerCase().replace(/^on/, '')] = value;
          }
          return true;
        },
      },
    );
  }

  registerLifecycle(name: string, callback: Callback) {
    if (typeof callback !== 'function') {
      return;
    }
    this.lifecycle[name] = this.lifecycle[name] || new Set([]);
    this.lifecycle[name].add(callback);

    return () => {
      this.lifecycle[name].delete(callback);
    };
  }
}

export const PageLifecycleGlobalContext = React.createContext<LifecycleContext | null>(null);
export const AppLifecycleGlobalContext = React.createContext<LifecycleContext | null>(null);
export const ComponentLifecycleGlobalContext = React.createContext<LifecycleContext | null>(null);
