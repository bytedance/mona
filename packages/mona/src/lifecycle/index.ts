export { usePageEvent, useAppEvent } from './hooks';
export * from './context';
