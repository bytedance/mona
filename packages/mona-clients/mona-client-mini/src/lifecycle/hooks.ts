import { useLayoutEffect, useContext } from 'react';
import { LifecycleContext, Callback, PageLifecycleGlobalContext } from './context';

// for app
export const appLifecycleContext = new LifecycleContext();
export function useAppEvent(eventName: string, callback: Callback) {
  useLayoutEffect(() => {
    appLifecycleContext.registerLifecycle(eventName, callback);
  });
}

// for page
export function usePageEvent(eventName: string, callback: Callback) {
  const pageLifecycleContext = useContext(PageLifecycleGlobalContext);

  useLayoutEffect(() => {
    if (pageLifecycleContext) {
      pageLifecycleContext.registerLifecycle(eventName, callback);
    }
  });
}
