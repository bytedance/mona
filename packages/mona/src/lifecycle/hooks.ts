import { useLayoutEffect, useContext } from 'react';
import { AppLifecycle, AppLifecycleGlobalContext, Callback, PageLifecycle, PageLifecycleGlobalContext } from './context';

// for app
export function useAppEvent(eventName: `${AppLifecycle}`, callback: Callback) {
  const appLifecycle = useContext(AppLifecycleGlobalContext);
  
  useLayoutEffect(() => {
    let clear = appLifecycle?.registerLifecycle(eventName, callback);
    return () => clear?.();
  }, [callback, eventName, appLifecycle]);
}

// for page
export function usePageEvent(eventName: `${PageLifecycle}`, callback: Callback) {
  const pageLifecycleContext = useContext(PageLifecycleGlobalContext);

  useLayoutEffect(() => {
    let clear = pageLifecycleContext?.registerLifecycle(eventName, callback);
    return () => clear?.();
  }, [callback, eventName, pageLifecycleContext]);
}
