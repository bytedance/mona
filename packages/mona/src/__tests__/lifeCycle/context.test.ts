import { LifecycleContext, PageLifecycle, AppLifecycle, ComponentLifecycle } from '../../lifecycle/context';

describe('lifeCycleName', () => {
  it('AppLifecycle SnapShot', () => {
    expect(AppLifecycle).toMatchSnapshot();
  });

  it('PageLifecycle SnapShot', () => {
    expect(PageLifecycle).toMatchSnapshot();
  });

  it('ComponentLifecycle SnapShot', () => {
    expect(ComponentLifecycle).toMatchSnapshot();
  });
});

describe('LifecycleContext', () => {
  let ctx: LifecycleContext;
  const testCb1 = () => {};
  const testCb2 = () => {};

  beforeEach(() => {
    ctx = new LifecycleContext();
  });
  it('donâ€™t  pass same cb', () => {
    // repeat
    ctx.registerLifecycle('load', () => {});
    ctx.registerLifecycle('onLoad', () => {});
    ctx.registerLifecycle('OnLoad', () => {});
    expect(Array.from(ctx.lifecycle[PageLifecycle.load].values()).length).toBe(3);
  });
  it('pass same cb', () => {
    // repeat
    ctx.registerLifecycle('load', testCb1);
    ctx.registerLifecycle('onLoad', testCb1);
    ctx.registerLifecycle('OnLoad', testCb1);
    expect(Array.from(ctx.lifecycle[PageLifecycle.load].values()).length).toBe(1);
  });
  it('clear', () => {
    // repeat
    const clear1 = ctx.registerLifecycle(PageLifecycle.load, testCb1);
    ctx.registerLifecycle(PageLifecycle.load, testCb1);
    const clear2 = ctx.registerLifecycle(PageLifecycle.load, testCb2);
    expect(Array.from(ctx.lifecycle[PageLifecycle.load].values()).length).toBe(2);
    expect(ctx.lifecycle[PageLifecycle.load].has(testCb2)).toBeTruthy();
    expect(ctx.lifecycle[PageLifecycle.load].has(testCb1)).toBeTruthy();

    // clear()
    clear1();
    expect(Array.from(ctx.lifecycle[PageLifecycle.load].values()).length).toBe(1);
    expect(ctx.lifecycle[PageLifecycle.load].has(testCb1)).toBeFalsy();
    expect(ctx.lifecycle[PageLifecycle.load].has(testCb2)).toBeTruthy();

    // clear2()
    clear2();
    expect(Array.from(ctx.lifecycle[PageLifecycle.load].values()).length).toBe(0);
  });
});
