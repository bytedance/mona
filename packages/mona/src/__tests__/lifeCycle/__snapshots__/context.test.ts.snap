// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lifeCycleName AppLifecycle SnapShot 1`] = `
Object {
  "error": "onError",
  "hide": "onHide",
  "launch": "onLaunch",
  "pageNotFound": "onPageNotFound",
  "show": "onShow",
  "unhandledRejection": "onUnhandledRejection",
}
`;

exports[`lifeCycleName ComponentLifecycle SnapShot 1`] = `
Object {
  "attached": "onAttached",
  "created": "onCreated",
  "detached": "onDetached",
  "moved": "onMoved",
  "ready": "onReady",
}
`;

exports[`lifeCycleName PageLifecycle SnapShot 1`] = `
Object {
  "hide": "onHide",
  "load": "onLoad",
  "pageScroll": "onPageScroll",
  "pullDownRefresh": "onPullDownRefresh",
  "reachBottom": "onReachBottom",
  "ready": "onReady",
  "resize": "onResize",
  "shareAppMessage": "onShareAppMessage",
  "show": "onShow",
  "unload": "onUnload",
}
`;
